src/main/java/com/arpnetworking/tsdcore/sinks/PeriodicStatisticsSink.java

```diff
@@ -240,47 +247,55 @@ public void run() {
+            return Objects.equals(_metricName, otherKey._metricName)
+                    && Objects.equals(_key, otherKey._key);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(_metricName, _key);
+        }
+
+        MetricKey(final String metricName, final Key key) {
+            _metricName = metricName;
+            _key = key;
+        }
+
+        private final String _metricName;
+        private final Key _key;
+    }
+
     private class Bucket {
 
         public void record(final PeriodicData periodicData, final long now) {
-            _aggregatedData.addAndGet(periodicData.getData().size());
-
-            // TODO(ville): Remove the period and statistic from the unique statistics identifier.
-            // TODO(ville): Convert the serialized identity to use a Set<Key> instead.
-            final String fqsnPrefix = new StringBuilder()
-                    .append(periodicData.getDimensions().getParameters().entrySet().stream()
-                            .map(entry -> entry.getKey() + "=" + entry.getValue())
-                            .collect(Collectors.joining(";")))
-                    .append(getPeriodAsString(periodicData.getPeriod()))
-                    .toString();
-            final String serviceMetricPrefix = new StringBuilder()
-                    .append(periodicData.getDimensions().getService()).append(".")
-                    .toString();
-
-            final Set<String> uniqueMetrics = Sets.newHashSet();
-            for (final Map.Entry<String, AggregatedData> entry : periodicData.getData().entries()) {
-                final String metricName = entry.getKey();
-                final AggregatedData datum = entry.getValue();
-                final String fqsn = new StringBuilder()
-                        .append(fqsnPrefix)
-                        .append(metricName).append(".")
-                        .append(datum.getStatistic().getName()).append(".")
-                        .toString();
-
-                final String serviceMetric = new StringBuilder()
-                        .append(serviceMetricPrefix)
-                        .append(metricName)
-                        .toString();
-
-                _uniqueMetrics.get().add(serviceMetric);
-
-                _uniqueStatistics.get().add(fqsn);
-
-                if (uniqueMetrics.add(serviceMetric)) {
-                    // Allow each service/metric in the periodic data to contribute
-                    // its population size (samples processed) to the sample count.
-                    _metricSamples.accumulate(datum.getPopulationSize());
-                }
+            final ImmutableMultimap<String, AggregatedData> data = periodicData.getData();
+            _aggregatedData.addAndGet(data.size());
+
+            for (final String metricName : data.keySet()) {
+                final AggregatedData firstDatum = data.get(metricName).iterator().next();
```

Comment:
I'm confused here how this is equivalent to the above. In the older code, we were iterating over all metricName, AggregatedData entries but here we're only taking the first one we see for a given metricname.

For `_uniqueMetrics` and `_uniqueStatistics` this would be equivalent sure, since the other AggregatedData would have the same set of dimensions but wouldn't `_metricSamples` be thrown off?

Reply:
Ah nevermind this comment, I see that the javadoc says that this should only take in periodicData for a single period, so this list would only have one element.

I think I would still rather see that precondition asserted instead of silently emitting incorrect metrics.